Fix dockerfile.security.missing-user rules

saghaulor opened the pull request:
- A `USER` directive can appear after the `CMD` or` ENTRYPOINT` directive and still be valid
- Updated sample dockerfiles with code comments
- Relates to #3438

saghaulor committed:
Fix dockerfile.security.missing-user rules

- A `USER` directive can appear after the `CMD` or` ENTRYPOINT` directive and still be valid
- Updated sample dockerfiles with code comments

0xDC0DE commented on dockerfile/security/missing-user-entrypoint.dockerfile:
```diff
@@ -9,6 +9,6 @@ RUN pip3 install semgrep
 
-# TODO: metavar bug
-# ok: missing-user-entrypoint
+# TODO: metavar ellipses bug, this should be a finding but is a false negative
+# ruleid: missing-user-entrypoint
```
For failing testcases that are related to the engine instead of the rule, we also have `todook` and `todoruleid`. This allows our tests to pass for now, and the rule in its current state to be published. But also allows our developers to verify what the intended behaviour is. If they make an engine update that fixes this problem, they will update the test syntax.
```suggestion
# TODO: metavar ellipses bug, this should be a finding but is a false negative
# todoruleid: missing-user-entrypoint
```

0xDC0DE commented on dockerfile/security/missing-user-entrypoint.fixed.dockerfile:
```diff
@@ -7,9 +7,10 @@ RUN git clone https://github.com/returntocorp/semgrep
 
-# TODO: metavar bug
+# TODO: metavar ellipses bug
 # ok: missing-user-entrypoint
 ENTRYPOINT ["semgrep", "--config", "localfile", "targets"]
```
```suggestion
# todoruleid: missing-user-entrypoint
ENTRYPOINT ["semgrep", "--config", "localfile", "targets"]
```

0xDC0DE commented on dockerfile/security/missing-user-entrypoint.yaml:
```diff
@@ -4,6 +4,7 @@ rules:
     - pattern: |
         ENTRYPOINT $...VARS
     - pattern-not-inside: |
+        ...
         USER $USER
         ...
```
Not sure if this can be helped, but starting a pattern with an ellipses is very slow! (I think we have CI checks to not allow this)

0xDC0DE replied:
To avoid a starting ellipses, maybe we could do something like this?

```
    - pattern: |
        ENTRYPOINT $...VARS
    - pattern-not-inside:
        USER $USER
        ...
        ENTRYPOINT $...VARS
    - pattern-not-inside:
        CMD $...VARS
        ...
        ENTRYPOINT $USER
```

0xDC0DE commented on dockerfile/security/missing-user.dockerfile:
```diff
@@ -6,12 +6,9 @@ FROM busybox
 
-# TODO: metavar ellipses bug
-# ok: missing-user
+# TODO: metavar ellipses bug, this should be a failure but is a false negative
+# ruleid: missing-user
```
```suggestion
# TODO: metavar ellipses bug, this should be a failure but is a false negative
# todoruleid: missing-user
```

0xDC0DE commented on dockerfile/security/missing-user.yaml:
```diff
@@ -4,6 +4,7 @@ rules:
     - pattern: |
         CMD $...VARS
     - pattern-not-inside: |
+        ...
         USER $USER
         ...
```
To avoid a starting ellipses, maybe we could do something like this?

```
    - pattern: |
        CMD $...VARS
    - pattern-not-inside:
        USER $USER
        ...
        CMD $...VARS
    - pattern-not-inside:
        CMD $...VARS
        ...
        USER $USER
```

0xDC0DE commented:
Thanks for your contributions, @saghaulor ! These are meaningful updates, let's try to get that CI to pass so we can merge this PR!

0xDC0DE committed:
Merge branch 'develop' into fix_dockerfile.security.missing-user

saghaulor commented:
@0xDC0DE thank you, I have returned from a long needed vacation, I'll try to make these changes soon.

saghaulor commented:
@0xDC0DE I finally had a chance to look into this more. It seems that this metavariable ellipsis bug is what is preventing us from writing an optimal rule. I tried to write a rule like you specified and it didn't work. The missing-user.fixed.dockerfile ends up in the results, which should not happen.

```bash
❯ head -n20 missing-user.yaml
rules:
- id: missing-user
  patterns:
    - pattern: |
        CMD $...VARS
    - pattern-not-inside: |
        USER $USER
        ...
        CMD $...VARS
    - pattern-not-inside: |
        CMD $...VARS
        ...
        USER $USER
  fix: |
    USER non-root
    CMD $...VARS
  message: By not specifying a USER, a program in the container may run as 'root'. This is a security
    hazard. If an attacker can control a process running as root, they may have control over the container.
    Ensure that the last USER in a Dockerfile is a USER other than 'root'.
  severity: ERROR

❯ semgrep scan --config=missing-user.yaml


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 14 files (only git-tracked) with 1 Code rule:

  CODE RULES
  Scanning 8 files.

  SUPPLY CHAIN RULES

  No rules to run.


  PROGRESS

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌─────────────────┐
│ 2 Code Findings │
└─────────────────┘

    missing-user.dockerfile
   ❯❯❱ missing-user
          By not specifying a USER, a program in the container may run as 'root'. This is a security hazard.
          If an attacker can control a process running as root, they may have control over the container.
          Ensure that the last USER in a Dockerfile is a USER other than 'root'.

           ▶▶┆ Autofix ▶ USER non-root CMD semgrep --config localfile targets
           10┆ CMD semgrep --config localfile targets

    missing-user.fixed.dockerfile
   ❯❯❱ missing-user
          By not specifying a USER, a program in the container may run as 'root'. This is a security hazard.
          If an attacker can control a process running as root, they may have control over the container.
          Ensure that the last USER in a Dockerfile is a USER other than 'root'.

           ▶▶┆ Autofix ▶ USER non-root CMD semgrep
           11┆ CMD semgrep
```

If we change the rule to remove the metavariable ellipsis behavior, it works as expected, namely, the missing-user.fixed.dockerfile is no longer reporting a finding. Please note that I only changed the second `pattern-not-inside` clause to remove the metavariable ellipsis behavior as that is what would trigger in the fixed missing-user.fixed.dockerfile.

```bash
❯ head -n20 missing-user.yaml
rules:
- id: missing-user
  patterns:
    - pattern: |
        CMD $...VARS
    - pattern-not-inside: |
        USER $USER
        ...
        CMD $...VARS
    - pattern-not-inside: |
        CMD ...
        ...
        USER $USER
  fix: |
    USER non-root
    CMD $...VARS
  message: By not specifying a USER, a program in the container may run as 'root'. This is a security
    hazard. If an attacker can control a process running as root, they may have control over the container.
    Ensure that the last USER in a Dockerfile is a USER other than 'root'.
  severity: ERROR

❯ semgrep scan --config=missing-user.yaml


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 14 files (only git-tracked) with 1 Code rule:

  CODE RULES
  Scanning 8 files.

  SUPPLY CHAIN RULES

  No rules to run.


  PROGRESS

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00


┌────────────────┐
│ 1 Code Finding │
└────────────────┘

    missing-user.dockerfile
   ❯❯❱ missing-user
          By not specifying a USER, a program in the container may run as 'root'. This is a security hazard.
          If an attacker can control a process running as root, they may have control over the container.
          Ensure that the last USER in a Dockerfile is a USER other than 'root'.

           ▶▶┆ Autofix ▶ USER non-root CMD semgrep --config localfile targets
           10┆ CMD semgrep --config localfile targets
```

This metavariable ellipsis bug also prevents the Semgrep engine from identifying incorrect patterns within the IDE.

The rule will trigger and highlight code when the Dockerfile specifies something like:

```dockerfile
CMD semgrep
```

But not when it's something like:

```dockerfile
CMD semgrep --severity=error
```

<img width="1351" alt="Screenshot 2024-09-13 at 1 32 28 PM" src="https://github.com/user-attachments/assets/bc904a6c-b0d3-49f1-bede-8a3c2ec64640">
<img width="527" alt="Screenshot 2024-09-13 at 1 33 25 PM" src="https://github.com/user-attachments/assets/95c4886f-e38e-43f7-abb7-57d9162c1c7f">

Therefore, I think we need to address the metavariable ellipsis bug before we can fix this rule for good. I believe that there is already an issue open here https://github.com/semgrep/semgrep/issues/9109 regarding this bug. I believe this issue is also related https://github.com/semgrep/semgrep/issues/9726

Please advise.

0xDC0DE committed:
Merge branch 'develop' into fix_dockerfile.security.missing-user
